### PR TITLE
Build front end

### PR DESCRIPTION
**Pull request by iMattPro**
A fully functioning front-end.
- [x] needs tests

Since there's no ACP yet, there's no data, but you can use this SQL query to insert some starter prefix to play with and test how this is working:

``` sql
INSERT INTO `phpbb_topic_prefixes` (`prefix_id`, `prefix_tag`, `prefix_enabled`, `prefix_parent_id`, `prefix_left_id`, `prefix_right_id`, `prefix_parents`, `forum_id`)
VALUES
    (1,'[BETA]',1,0,1,2,'',2),
    (2,'[3.2.x] [BETA]',1,0,3,4,'',2),
    (3,'[ABD]',1,0,5,6,'',2);
```

Seems like a big PR, but most of the new files are just unit tests.

### CODE REVIEW

**Comment by rxu** [comment]
Works good for me, including in combination with #4.

**Comment by iMattPro** [comment]
@michaelcullum

File: event/listener.php [hunk 1]
```diff
@@ -18,14 +21,147 @@
...
+
+	/**
+	 * Prepare topic prefix data for post submission
+	 *
+	 * @param \phpbb\event\data $event Event data object
+	 * @return null
+	 */
+	public function submit_prefix_data($event)
+	{
+		if (!($selected = $this->request->variable('topic_prefix', 0)))
```
Comment (michaelcullum) [inline event/listener.php:100]: Can this be split into two lines?

Comment (iMattPro) [inline event/listener.php:100]: I don't see any reason to, I kind of like it (i like it when the first bit of code is a return condition)

**Pull request merged**